### PR TITLE
Specify need to use *distributed* revision control

### PR DESCRIPTION
**Pull request by aepstein**
Traditional revision control forces a hierarchy on the project that interferes with collaborative development and product sharing.  The "pull request" approach to development can only happen with a distributed tool that allows each user to commit independently and decide on their own terms/timeline when and how to merge in others' work.

### CODE REVIEW

File: _plays/04.md [hunk 1]
```diff
@@ -12,7 +12,7 @@ We should use an incremental, fast-paced style of software development to reduce
 4. Keep delivery teams small and focused; limit organizational layers that separate these teams from the business owners
 5. Release features and improvements multiple times each month
 6. Create a prioritized list of features and bugs, also known as the “feature backlog” and “bug backlog”
-7. Use a source code version control system
+7. Use a source code distributed version control system
```
Comment (paultag) [inline _plays/04.md]: (Not actually commenting on the substance of this PR; I don't know that I have an opinion, so long as it's properly reviewed and tooled), but I think the phrasing:

> 7. Use a distributed source code version control system

would read a bit less awkwardly

Comment (cew821) [inline _plays/04.md]: I agree. @aepstein could you amend your pull request with one of these options?

**Comment by cew821** [comment]
Thanks for the suggestion. @paultag @krachmeleromb @alex @barkimedes @mdickers47 @MarinaMartin - I would love some thoughts on this proposal.

**Comment by alex** [comment]
I'm pro-DVCS in general; however it is _possible_ to do a pull-request workflow without it (e.g. Phabricator enables this with `svn`). I could go either way on inclusion of this in the playbook, but I think I fall at `+0` (weakly in favor).

**Comment by krachmeleromb** [comment]
I think that we are conflating two different concepts, branching/merging and centralized repository.  The definition of a distributed version control system is that there is no centralized repo: users all have working copies and exchange patches.  You can do a pull-request model without that, i.e. have a coherent branching mechanism.

**Comment by cew821** [comment]
@krachmeleromb Thanks for the clarification. The proposed change is from `Use a source code version control system` to `Use a distributed version control system for source code` - I would love to know if there is a consensus yes/no on this proposal from the engineering community in USDS.

**Comment by krachmeleromb** [comment]
@cew821 My own thought is that requiring "distributed" is overly prescriptive but I will ask the eng community for their opinion.

**Comment by aepstein** [comment]
There is a decent wikipedia article on [distributed version control](https://en.wikipedia.org/wiki/Distributed_version_control), that lists these advantages:

- Allows users to work productively when not connected to a network.
- Makes most operations much faster.
- Allows participation in projects without requiring permissions from project authorities, and thus arguably better fosters culture of meritocracy instead of requiring "committer" status.
- Allows private work, so users can use their changes even for early drafts they do not want to publish.
- Avoids relying on one physical machine as a single point of failure.
- Permits centralized control of the "release version" of the project
- On FLOSS software projects it is much easier to create a project fork from a project that is stalled because of leadership conflicts or design disagreements.

The article cites a [post](http://joelonsoftware.com/items/2010/03/17.html) which says in relevant part:

> With distributed version control, merges are easy and work fine. So you can actually have a stable branch and a development branch, or create long-lived branches for your QA team where they test things before deployment, or you can create short-lived branches to try out new ideas and see how they work.
> 
> This is too important to miss out on. This is possibly the biggest advance in software development technology in the ten years I’ve been writing articles here.

**Comment by robbiethegeek** [comment]
I think that we have code in version control is a huge step forward and doesn't prescribe a specific solution, would hate to force something that would create a contentious interaction / relationship because we forced a project to adopt yet another version control solution above and beyond one that was already in place that wasn't distributed.

**Comment by spoon16** [comment]
I am generally in favor of constraints that reduce the number of options in front of a decision maker. The proposed constraint massively reduces the options available when a VCS is being selected for a project.

However, I am only weakly in favor of this change. I believe the DVCS is implicitly covered by [Choose a modern technology stack](https://playbook.cio.gov/#play8).

**Comment by lorimt** [comment]
I'm inclined to agree with @robbiethegeek - my impression is the playbook defines what you *need* to succeed, not what's optimal, which I think is important for agencies who are working to reach the minimal bar, as well as agencies that have an OK if not perfect solution and might be better off spending energy elsewhere.  

The model I would *like* agencies to use is:
- "problems to be solved" (make it easier to find and reuse code, collaborate, etc)
- "features to look for" (central repository, replication, etc)
- "acceptable implementations" (OK options, good options)

The advantage being we can be less prescriptive about specific implementations as long as the underlying problems are solved.  The risk is someone convinces themselves a bad solution is acceptable despite not actually solving problems.

On the other hand, the playbook is much more clear and concise, which makes it much easier to communicate and push for.

**Comment by mbellotti** [comment]
I agree with @alex and @robbiethegeek and will add one extra note for consideration: Team Foundation Server claims to be distributed, but it only is if you swap out it's native VCS for git, a fact that may or may not be disclosed and/or understood. I feel like such a recommendation could push us into "private cloud" and "hybrid cloud" territory pretty easily.

Also since DVCS has only one major player (sorry Mercurial fans) I am also concerned that this could give ammunition to the "USDS is taking business away from the private sector" camp

**Comment by dnesting** [comment]
I am weakly against this proposal because I feel it is too prescriptive.  As others have said, overly prescriptive guidance is far more likely to become less relevant or appropriate over time.  But it is also possible that we have scenarios today where DVCS is the wrong solution, and I don't want people trying to push an inappropriate solution onto a problem by appealing to the playbook.  Using VCS within government projects is already ridiculously rare and I'd rather we just get *something* in use instead of litigating us all the way to git.

**Comment by cew821** [comment]
Based on the above, it does not look like we have a consensus agreement in favor of this change. Thank you anyways to @aepstein for the proposal, we really appreciate your interest and ideas. Please keep proposals for improving this document coming! :us:

**Pull request closed without merging**